src/main/java/eu/europeana/metis/sandbox/service/workflow/HarvestServiceImpl.java

```diff
@@ -82,77 +83,92 @@ public void harvestOaiPmh(String datasetId, Record.RecordBuilder recordDataEncap
 
       recordHeaderIterator.forEach(recordHeader -> {
-        OaiHarvestData completeOaiHarvestData = new OaiHarvestData(oaiHarvestData.getUrl(),
-            oaiHarvestData.getSetspec(),
-            oaiHarvestData.getMetadataformat(),
-            recordHeader.getOaiIdentifier());
-
-        if (recordHeader.isDeleted()) {
-          return ReportingIteration.IterationResult.CONTINUE;
-        }
-
-        currentNumberOfIterations.getAndIncrement();
-
-        if (currentNumberOfIterations.get() > maxRecords) {
-          datasetService.setRecordLimitExceeded(datasetId);
-          currentNumberOfIterations.set(maxRecords);
-          return ReportingIteration.IterationResult.TERMINATE;
-        }
-
-        recordInfoList.add(harvestOaiRecords(datasetId, completeOaiHarvestData,
-            recordDataEncapsulated));
-
-        return ReportingIteration.IterationResult.CONTINUE;
-      });
+            try {
+              OaiHarvestData completeOaiHarvestData = new OaiHarvestData(oaiHarvestData.getUrl(),
+                  oaiHarvestData.getSetspec(),
+                  oaiHarvestData.getMetadataformat(),
+                  recordHeader.getOaiIdentifier());
+
+              if (recordHeader.isDeleted()) {
+                return ReportingIteration.IterationResult.CONTINUE;
+              }
+
+              numberOfIterations.getAndIncrement();
+
+              if (numberOfIterations.get() > maxRecords) {
+                datasetService.setRecordLimitExceeded(datasetId);
+                numberOfIterations.set(maxRecords);
+                return ReportingIteration.IterationResult.TERMINATE;
+              }
+
+              recordInfoList.add(harvestOaiRecords(datasetId, completeOaiHarvestData, recordDataEncapsulated, recordInfoList));
+
+            } catch (RuntimeException harvestException) {
+              saveErrorWhileHarvesting(recordDataEncapsulated, recordHeader.getOaiIdentifier(), Step.HARVEST_OAI_PMH,
+                  harvestException);
+            }
+            return ReportingIteration.IterationResult.CONTINUE;
+          }
+      );
 
-      datasetService.updateNumberOfTotalRecord(datasetId, currentNumberOfIterations.get());
+      datasetService.updateNumberOfTotalRecord(datasetId, numberOfIterations.get());
 
     } catch (HarvesterException | IOException e) {
       throw new ServiceException("Error harvesting OAI-PMH records ", e);
     }
-    return recordInfoList;
-  }
 
-  private RecordInfo harvestOaiRecords(String datasetId, OaiHarvestData oaiHarvestData,
-      Record.RecordBuilder recordToHarvest) {
+    return recordInfoList.stream()
+                         .filter(Optional::isPresent)
+                         .map(Optional::get)
+                         .collect(Collectors.toList());
+  }
 
+  private Optional<RecordInfo> harvestOaiRecords(String datasetId, OaiHarvestData oaiHarvestData,
+      Record.RecordBuilder recordToHarvest,
+      List<Optional<RecordInfo>> optionalList) {
```

Comment:
I wonder what wants to be achieved by having this List with optionals, feels like an overuse of it, there is a nice article [here](Maybe there is even an overuse of optional, but following some recommendations here https://blogs.oracle.com/javamagazine/post/the-java-optional-class-11-more-recipes-for-preventing-null-pointer-exceptions#:~:text=Recipe%2015%3A%20Optional%20should%20not%20be%20used%20in%20the%20method%E2%80%99s%20arguments) that mentions doing this as not recommended
Also they have this other [one](https://blogs.oracle.com/javamagazine/post/12-recipes-for-using-the-optional-class-as-its-meant-to-be-used)
Both nice readings about the use of Optional class, of course they are not rules, but I will challenge if doing this is necessary

Reply:
@stzanakis correct. renamed to ```optionalRecordInfoList```
@acerdas84 thanks for the articles. Yes it gave me more insight about the use of optionals. In this case the intention of using optional is as a representation of the absence of value. I chose this path instead of going over null value.

Reply:
Not sure to understand fully but if you want to take out nulls can you just do something like
```java
list.filter(Objects::nonNull).collect(Collectors.toList());
```

Reply:
correct. I didn't want it to do it that way. that is why I choose Optional.

Reply:
Then I think this [one](https://blogs.oracle.com/javamagazine/post/12-recipes-for-using-the-optional-class-as-its-meant-to-be-used#:~:text=Recipe%2012%3A%20Don%E2%80%99t%20overuse%20Optional.) applies here. The overhead of creating one more level of indirection and a new object is not worth for such simple check, checking for null seems more appropriate, but what does everyone else thinks? maybe @stzanakis ?

Reply:
After also discussed with @stzanakis, @acerdas84 the optional use was removed and change to use nulls.